feat: Add ui.resolve to support URIs in dh.ui (#1215)
deephaven/deephaven-plugins: plugins/ui/docs/components/uri.md

# URI

> URIs are a way to reference Deephaven resources, such as tables or figures, from another instance. Deephaven UI has its own `ui.resolve` method that does not require any server to server communication. Instead, the web client will communicate directly with the appropriate server to get the resource.

Reviewer: ```suggestion
URIs are a way to reference Deephaven resources, such as tables or figures, from another instance. Deephaven UI has its own `ui.resolve` method that does not require server-to-server communication. Instead, the web client communicates directly with the appropriate server to get the resource.
```

## Usage

> Deephaven UI provides a `resolve` method (not to be confused with `resolve` method from the Deephaven URI package) that allows you to reference Deephaven resources from other instances. Unlike the Deephaven URI package, `ui.resolve` does not resolve the URI to its resource on the server, so you cannot apply operations to the resource.

Reviewer: ```suggestion
Deephaven UI provides a `resolve` method (not to be confused with the `resolve` method from the Deephaven URI package) that allows you to reference Deephaven resources from other instances. Unlike the Deephaven URI package, `ui.resolve` does not resolve the URI to its resource on the server, so you cannot apply operations to the resource.
```

Reviewer: add a link to the other resolve method?

> [!NOTE]

> > The only valid URIs for deephaven UI at the moment are for Deephaven Enterprise persistent queries.

Reviewer: ```suggestion
> Currently, the only valid URIs for Deephaven UI are for Deephaven Enterprise Persistent Queries.
```

> > See the [Deephaven Enterprise documentation](/enterprise/docs/deephaven-database/remote-tables-python/#uris) for more information on persistent query URIs. The optional parameters are ignored by `ui.resolve`.

Reviewer: ```suggestion
> See the [Deephaven Enterprise documentation](/enterprise/docs/deephaven-database/remote-tables-python/#uris) for more information on Persistent Query URIs. The optional parameters are ignored by `ui.resolve`.
```

### Plain references

> One way to use `ui.resolve` is to assign the reference to a variable which will be opened by the web UI just like you created the resource in the first place. This can be useful if you want to display tables from multiple sources in a single dashboard.

Reviewer: ```suggestion
One way to use `ui.resolve` is to assign the reference to a variable, which the web UI will open just as if you created the resource. This can be useful if you want to display tables from multiple sources in a single dashboard.
```

```py order=null
from deephaven import ui

t = ui.resolve("pq://MyPersistentQuery/scope/table") # Can't do t.update() or any other operations
p = ui.resolve("pq://MyPersistentQuery/scope/plot")

@ui.component
def basic_dashboard():
  return ui.panel(ui.flex(t, p), title="Table and Plot")

my_dashboard = ui.dashboard(basic_dashboard())
```

### Usage in UI components

> Some Deephaven UI components that accept tables as sources can also accept URIs. This includes [`ui.table`](table.md) and any components that accept an `item_table_source`. When using a URI with UI components, you can often just use the string without needing to call `ui.resolve`. However, if a component may take a string as a valid child (e.g., `ui.picker`), then you must use `ui.resolve` to distinguish between a string and a URI. You can always use `ui.resolve` in the places you can use just the string if you prefer to be explicit.

Reviewer: ```suggestion
Some Deephaven UI components that accept tables as sources can also accept URIs. This includes [`ui.table`](table.md) and any components that accept an `item_table_source`. When using a URI with UI components, you can often just use the string without needing to call `ui.resolve`. However, if a component may take a string as a valid child (e.g., `ui.picker`), then you must use `ui.resolve` to distinguish between a string and a URI. You can always use `ui.resolve` in places where you can use just the string if you prefer to be explicit.
```
